### DISCUSSION TITLE
Add header option parameter while using JWTAuthentication to generate jwt token

### DISCUSSION THREAD

**Discussion by DanielChuDC** [fastapi-users/fastapi-users#756] [Ideas]
Would like to add optional header parameters while generating for JWT.

```
jwt_authentication = JWTAuthentication(secret=settings.JWT_SECRET_KEY,
                                       lifetime_seconds=settings.JWT_ACCESS_TOKEN_EXPIRE_MINUTES * 60,
                                       tokenUrl="auth/jwt/login")
```

to 

```
jwt_authentication = JWTAuthentication(secret=settings.JWT_SECRET_KEY,
                                       lifetime_seconds=settings.JWT_ACCESS_TOKEN_EXPIRE_MINUTES * 60,
                                       tokenUrl="auth/jwt/login",
                                       headers={"kid": "230498151c214b788dd97f22b85410a5"})
```

**Comment by frankie567** [comment]
This is something you can do by overloading the `_generate_token` method. For example:

```py
from fastapi_users.authentication import JWTAuthentication
from fastapi_users.jwt import generate_jwt

class CustomJWTAuthentication(JWTAuthentication):
    async def _generate_token(self, user) -> str:
        data = {"user_id": str(user.id), "aud": self.token_audience, "kid": "MY_VALUE"}
        return generate_jwt(data, self.secret, self.lifetime_seconds)
```

**Reply by DanielChuDC** [reply]
thanks for replying! @frankie567 . I did the overloading and yea it works :) thanks for your help.

Another thought is how do we implement the RS256/ES256 encryption with this library?

**Reply by frankie567** [reply]
The JWT is already hashed using the HS256 algorithm: https://github.com/fastapi-users/fastapi-users/blob/7ae2042500671552bdf134639abb29dfb05ff2ca/fastapi_users/jwt.py

**Reply by DanielChuDC** [reply]
Right, what if I would like to upgrade to RS256?

**Reply by frankie567** [reply]
Well, in this case you would have to implement your authentication backend yourself to provide both public and private keys in the constructor and use appropriate JWT [encode / decode function calls](https://pyjwt.readthedocs.io/en/stable/usage.html#encoding-decoding-tokens-with-rs256-rsa).

It should be fairly similar to the built-in one: https://github.com/fastapi-users/fastapi-users/blob/7ae2042500671552bdf134639abb29dfb05ff2ca/fastapi_users/authentication/jwt.py

**Reply by DanielChuDC** [reply]
Alright! thank you! @frankie567
